### PR TITLE
Using go routines for gameserver creation and deletion

### LINKED ISSUE #213: Asynchronous calls when creating and deleting gameservers in batches
Currently the controller only creates or deletes a max of 20 gameservers synchronously in each reconcile loop to avoid the cycle getting too long. To further improve this it could be useful to make this calls asynchronously.

### PR DESCRIPTION
**Pull request by javier-op**
Fixes #213 
The reconcile function still has a maximum of 20 creations and deletions per loop, because having small loops show better performance, but now those calls are made asynchronously with go routines.

### CODE REVIEW

File: pkg/operator/controllers/gameserverbuild_controller.go [hunk 1]
```diff
@@ -63,6 +63,11 @@ const (
 	maxNumberOfGameServersToDelete = 20
 )
 
+type MutexMap struct {
```
Comment (dgkanatsios) [inline pkg/operator/controllers/gameserverbuild_controller.go:66]: can we add a comment that it's used for the creations/deletions expectations code?

File: pkg/operator/controllers/gameserverbuild_controller.go [hunk 2]
```diff
@@ -359,23 +384,34 @@ func (r *GameServerBuildReconciler) deleteNonActiveGameServers(ctx context.Conte
...
-	deletedGameServersCount := 0
-	for i := 0; i < len(gameServers.Items) && deletedGameServersCount < totalNumberOfGameServersToDelete; i++ {
+	// an error channel for the go routines to write errors
+	errCh := make(chan error, totalNumberOfGameServersToDelete)
+	// a waitgroup for async deletion calls
+	var wg sync.WaitGroup
+	deletionCalls := 0
+	for i := 0; i < len(gameServers.Items) && deletionCalls < totalNumberOfGameServersToDelete; i++ {
 		gs := gameServers.Items[i]
-		// we're deleting only initializing/pending/standingBy servers, never touching active
```
Comment (dgkanatsios) [inline pkg/operator/controllers/gameserverbuild_controller.go:365]: was this comment moved somewhere?

File: pkg/operator/controllers/gameserverbuild_controller.go [hunk 3]
```diff
@@ -359,23 +384,34 @@ func (r *GameServerBuildReconciler) deleteNonActiveGameServers(ctx context.Conte
 	gsb *mpsv1alpha1.GameServerBuild,
 	gameServers *mpsv1alpha1.GameServerList,
 	totalNumberOfGameServersToDelete int) error {
-	deletedGameServersCount := 0
-	for i := 0; i < len(gameServers.Items) && deletedGameServersCount < totalNumberOfGameServersToDelete; i++ {
+	// an error channel for the go routines to write errors
+	errCh := make(chan error, totalNumberOfGameServersToDelete)
+	// a waitgroup for async deletion calls
+	var wg sync.WaitGroup
+	deletionCalls := 0
```
Comment (dgkanatsios) [inline pkg/operator/controllers/gameserverbuild_controller.go:391]: is this being used?

File: pkg/operator/controllers/gameserverbuild_controller.go (hunk 1 above)
Comment (javier-op) [reply pkg/operator/controllers/gameserverbuild_controller.go:68]: Done!

File: pkg/operator/controllers/gameserverbuild_controller.go (hunk 2 above)
Comment (javier-op) [reply pkg/operator/controllers/gameserverbuild_controller.go:365]: I deleted it by accident, now it's back haha

**Review by dgkanatsios — APPROVED** [review]

**Pull request merged**